Design/histogram_type.md

```diff
@@ -0,0 +1,251 @@
+   They can be represented as variances or standard deviations.
+   We thus introduce 8 types: `BinEdgeVariances`, `BinEdgeStandardDeviations`, `PointVariances`, `PointStandardDeviations`, `CountVariances`, `CountStandardDeviation`, `FrequencyVariances`, `FrequencyStandardDeviations`.
+
+The amount of new types introduced may seem appalling at first, but we usually do not have to deal with many of them:
+
+1. For 'standard' operations you do not need to work with the low-level types, but can use the interface of `Histogram`, e.g.,
+
+   ```cpp
+   Histogram hist1;
+   Histogram hist2;
+   // Initialize histograms
+   // ...
+   // Adding histograms can automatically add `Y`, and propagates the errors properly
+   auto result = hist1 + hist2;
+   ```
+
+   Details of the `Histogram` interface are still open, but it should be simple to extend it once the roll-out is done.
+
+2. `BinEdges` is convertible to `Points`. We can thus avoid the currently prevalent way of dealing with this:
+
+   ```cpp
+   // Old:
+   const bool isHistogram = ws->isHistogramData();
+   for (size_t i = 0; i < Y.size(); ++i) {
+     x = isHistogram ? (0.5 * (X[i] + X[i + 1])) : X[i];
+     // use x
+   }
+   
+   // New: If you need points, not bin-edges, just read them, Histogram can convert
+   const auto &points = histogram.points();
+   for (size_t i = 0; i < Y.size(); ++i) {
+     // use points[i]
+   }
+   ```
+
+   The reverse conversion is of course not unique, so we will still have algorithms that refuse to work with histograms that store point data.
+
+3. `Counts` and `Frequencies` can be converted into each other (internally in the `Histogram` type).
+   When implementing an algorithm you can thus simply choose which representation you prefer to work with, e.g.,
+
+   ```cpp
+   Counts counts { 16, 56758, 34, 0 };
+   histogram.setCounts(counts);
+   //...
+   // We set counts, but can read frequencies somewhere else
+   const auto &frequencies = histogram.frequencies();
+   // Errors are dealt with implicitly unless specified otherwise
+   const auto &errors = histogram.countStandardDeviations();
+   // errors[0] should now be 4
+   ```
+
+4. As shown in the previous example, in many cases you do not have to deal with errors yourself.
+   As long as the standard deviation is simply the square root of the counts it will be computed and propagated under the hood.
+
+5. `Variances` and `StandardDeviations` can be automatically converted, you do not have to deal with it yourself.
+
+
+#### Details
+
+- The different low-level types have been introduced in the previous section.
+  To avoid highly duplicate code and tests, such as for the four `Variance` types, these are implemented using the curiously recurring template pattern (CRTP) which provides something like "mixins".
+- Jon pointed out that it may make most sense to default to storing data as `Frequecies`, as OpenGenie did by default.
```

Comment:
@AndreiSavici I used the term 'frequency' in a more general interpretation here, it could be, e.g., spatial frequency, angular frequency, or in you example 'thermal frequency' (counts/Kelvin). So basically `Frequency` would always refer to `Counts` divided by `BinWidth`, and the interpretation of `Frequency` would follow from the meaning of the `BinWidth`, i.e., the `X` unit.

I am open to renaming `Frequency` if you think in can cause trouble. @OwenArnold @martyngigg, what do you think?